### ISSUE TITLE
Cannot access property $… on … \Post|null

### ISSUE THREAD

**Issue by mfn** [nunomaduro/larastan#611]
- Larastan Version: 0.6.1
- `--level` used: max (8)

### Description
In https://github.com/rebing/graphql-laravel/pull/655 I'm trying to upgrade to larastan 0.6.1 but I'm facing with hundreds of errors like these:
```
…
 ------ --------------------------------------------------------------------------------------
  Line   tests/Database/SelectFields/NestedRelationLoadingTests/NestedRelationLoadingTest.php
 ------ --------------------------------------------------------------------------------------
  78     Cannot access property $body on Rebing\GraphQL\Tests\Support\Models\Post|null.
  79     Cannot access property $id on Rebing\GraphQL\Tests\Support\Models\Post|null.
…
```
See also https://github.com/rebing/graphql-laravel/pull/655/checks?check_run_id=851495801#step:7:20

Here starts the test in question https://github.com/rebing/graphql-laravel/blob/4068c15911822581af33ce60113387a372fa7783/tests/Database/SelectFields/NestedRelationLoadingTests/NestedRelationLoadingTest.php#L17

The particular line of code (line 78):
```php
                                'body' => $users[0]->posts[0]->body,
```
At the beginning of this test, `$users` is type-hinted as this:
```php
        /** @var User[] $users */
        $users = factory(User::class, 2)
            ->create()
            ->each(function (User $user): void {
                factory(Post::class, 2)
                    ->create([
                        'user_id' => $user->id,
                    ])
                    ->each(function (Post $post): void {
                        factory(Comment::class, 2)
                            ->create([
                                'post_id' => $post->id,
                            ]);
                    });
            });
```
When larastan/phpstan resolves the access to `$users[0]->posts[0]` it seems it trips of this definition in the `User`-model https://github.com/rebing/graphql-laravel/blob/4068c15911822581af33ce60113387a372fa7783/tests/Support/Models/User.php#L13
```php
 * @property-read \Illuminate\Database\Eloquent\Collection|Post[] $posts
```
If I change this to
```php
 * @property-read Post[] $posts
```
it reduces the errors by a few hundreds and further adapting similar models like this makes them completely go away.

Technically I don't need them that way however I was simply "emulating" how https://github.com/barryvdh/laravel-ide-helper usually generates them.

Basically the thing **I** don't understand is: why would the use if `\Illuminate\Database\Eloquent\Collection` actually may think the analyzer it's `|null`-able? At least to me this is how it looks.

Thanks!

**Comment by Daanra** [comment]
Might be a PHPStan issue?

Seems like this is the core of the problem: https://phpstan.org/r/ddcb9c47-f411-4909-bbe8-bdc46ca6e79c

Couldn't find any related issues on their repository but this seems like undesired behavior to me.

**Comment by canvural** [comment]
Larastan converts docblocks like `\Illuminate\Database\Eloquent\Collection|Post[]` to `\Illuminate\Database\Eloquent\Collection<Post>`. So this explains why the errors are gone when you change it to just `Post[]` And by combining this by the fact @Daanra pointed out, this might be PHPStan issue.

**Comment by canvural** [comment]
I think it's coming from here: https://github.com/phpstan/phpstan-src/blob/master/stubs/ArrayObject.stub#L18

I don't know why this is explicitly specified. Maybe in some scenarios, it returns `null` Does anybody know? cc @ondrejmirtes 

Psalm also [has it](https://github.com/vimeo/psalm/blob/master/src/Psalm/Internal/Stubs/CoreGenericClasses.phpstub#L285) as `TValue|null` but ignores the `null` return with an annotation.

But it is possible to override `offsetGet` in our Eloquent Collection stub. Just don't know if it is a good idea.

**Comment by szepeviktor** [comment]
> Does anybody know?

A commenter talks about this: https://www.php.net/manual/en/arrayaccess.offsetget.php#124324

**Comment by canvural** [comment]
> > Does anybody know?
> 
> A commenter talks about this: [php.net/manual/en/arrayaccess.offsetget.php#124324](https://www.php.net/manual/en/arrayaccess.offsetget.php#124324)

I saw that. But I think that would result in `Notice: Undefined offset` error, not returning `null`

**Comment by szepeviktor** [comment]
https://github.com/laravel/framework/blob/67d6d9688122dcf6c449b74ddc873fc9934e81e6/src/Illuminate/Support/Collection.php#L1332-L1352

AFAIK this shows that Laravel never returns null.

**Comment by ondrejmirtes** [comment]
You're actually getting null this way, so PHPStan is correct: https://3v4l.org/gCkVf PHPstan cannot know if that offset exists or not.

**Comment by mfn** [comment]
Wow, thanks for all the feedback!

I guess I shouldn't have posted this so late as it's kind of "obvious" if you think about it, `$user[0]->posts[<any number not hitting anything]` surely _could_ return `null`.

Probably the tests are not well designed and it shouldn't be done that way. But since I "know" the factory setup, it's easier to express the assertions that way. Probably the https://wiki.php.net/rfc/nullsafe_operator could help here, if we get it.

So seems to me, without having to refactor the tests (ideas welcome but I don't expect anyone here doing my work 😏 ) I should simply ignore them.

**Comment by canvural** [comment]
I think you can also do `$user->first()->posts->first()->body`

**Comment by mfn** [comment]
It doesn't really seem to improve things; I changed the typehint to:
```php
        /** @var Collection|User[] $users */
        $users = factory(User::class, 2)
```
but still
- `  75     Cannot access property $id on Rebing\GraphQL\Tests\Support\Models\User|null.`
  => line 75: `id' => (string) $users->first()->id`
- or this
  ```
  79     Cannot access property $body on Rebing\GraphQL\Tests\Support\Models\Post|null.
  79     Cannot access property $posts on Rebing\GraphQL\Tests\Support\Models\User|null.
  ```
  for `'body' => $users->first()->posts->first()->body`

Additionally using the method breaks IDE (PhpStorm) understanding where things lead to, something which is really useful for such kind of tests.

**Comment by mfn** [comment]
For now I decided to simply ignore them as don't find the alternatives compelling and currently favour the support I get from the IDE directly.

Thanks for all the insights 🤗
